rfcs/20251203-grouped-gemm-support/README.md

```diff
@@ -0,0 +1,604 @@
+offsets = [0, M0, M0+M1, M0+M1+M2, M0+M1+M2+M3]
+
+Note: Why "sparse"? M_i can be zero if an expert receives no tokens.
+In that case, the next expert starts at the same offset, similar to CSR format row pointers.
+Example: if M1 = 0, then offsets = [0, M0, M0, M0 + M2, M0 + M2 + M3]
+```
+
+The grouped memory descriptor defines:
+- Block count: `num_experts` (known at creation)
+- Total elements: `num_input_tokens x TOP_K x K` (known at creation)
+- Block dimensions: `[Mi x K]` where `Mi` is runtime, `K` is fixed
+- Data types: values and offsets (known at creation)
+- Values buffer: Contiguous data sum(Mi x K) (provided at execution)
+- Offsets buffer: `[0, M0, M0 + M1, ...]` size `num_experts + 1` (at execution)
+
+This representation is analogous to CSR sparse format. CSR uses row
+pointers and column indices to describe sparse structure. The grouped
+descriptor uses offsets to mark block boundaries.
+Empty blocks (experts with zero tokens) are skipped via offsets.
+Structure is known at primitive creation, while runtime dimensions and buffers
+are resolved at execution time using device-side data.
+
+#### Memory Descriptor API
+
+The grouped memory descriptor is created similarly to CSR descriptors:
+
+```cpp
+static memory::desc grouped(
+    const dims &adims,               // Overall tensor dims [total_M, K]
+    data_type adata_type,            // Elements data type
+    dim group_count,                 // Group size or number of experts (all, not just active ones)
+    const dims &grouped_dims,        // Group dimensions (clarification below)
+    data_type offsets_dt = s32,      // Offset buffer dtype (default int32)
+    const dims &astrides = {});      // Optional strides (default row-major)
+```
+
+Suggestion is to allow specifying `grouped_dims` in two ways for convenience:
+- **Shared dimensions**: `{M, K}` where M can be RUNTIME_DIM_VAL (size = 2)
+    - Useful for device-side runtime dimension M
+- **Per-group dimensions**: `{M0, K, M1, K, ...}` (size = group_count * 2)
+    - Useful when all dimensions are known on the host in advance
+
+The descriptor specifies a memory object with 2 buffers:
+- Buffer 0: Values, contiguous data organized as groups
+- Buffer 1: Offsets, cumulative row offsets of size `group_count + 1`
+
+Example:
+```cpp
+auto src_md = memory::desc::grouped(
+    {total_tokens, K},              // Overall shape
+    dt::f32,                        // Data type
+    num_experts,                    // Group count
+    {DNNL_RUNTIME_DIM_VAL, K},      // Shared per-group dims
+    dt::s32);                       // Offset dtype
+```
+
+#### Memory Object Creation
+
+Memory objects are created with 2 buffer pointers:
+
+```cpp
+// Buffers are coming from Framework side (could be device memory)
+float* values;       // total_tokens * K contiguous data
+int32_t* offsets;    // num_experts + 1 offsets
+
+// Create memory object with 2 buffers
+memory src_mem(src_md, eng, {values, offsets});
```

Comment:
It's just got to me that with this approach a destination memory object demands two pointers, where `values` would be used as output to store values and `offsets` would be used as input to determine how to distribute those `values` unless I'm missing something. If this is the case, it'll be a unique outstanding behavior besides the fact that pointers for dst memory are considered non-const, thus, may mean that offsets might be changed by the library.

Reply:
Yeah, that's correct understanding, you could see dst mem creation and usage right below in the "Primitive Execution" section for instance. Offsets are not to be changed by the library, they are filled and passed from the user's side in order to describe/resolve the memory.

Is there any particular issue that you see or is the point just to make sure that we clarify about offsets being input parameter?